sys/usb/usbus/dfu/dfu.c

```diff
@@ -0,0 +1,260 @@
+/*
+ * Copyright (C) 2020 Mesotic SAS
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup usbus_dfu
+ * @{
+ * @file USBUS implementation for device firmware upgrade
+ * 
+ *
+ * @author  Dylan Laduranty <dylan.laduranty@mesotic.com>
+ * @}
+ */
+
+#define USB_H_USER_IS_RIOT_INTERNAL
+
+#include "usb/dfu.h"
+#include "usb/descriptor.h"
+#include "usb/usbus.h"
+#include "usb/usbus/control.h"
+
+#include "periph/flashpage.h"
+#include "periph/pm.h"
+
+#include "riotboot/slot.h"
+
+#include <string.h>
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+static void _event_handler(usbus_t *usbus, usbus_handler_t *handler,
+                          usbus_event_usb_t event);
+static int _control_handler(usbus_t *usbus, usbus_handler_t *handler,
+                            usbus_control_request_state_t state,
+                            usb_setup_t *setup);
+static void _transfer_handler(usbus_t *usbus, usbus_handler_t *handler,
+                              usbdev_ep_t *ep, usbus_event_transfer_t event);
+static void _init(usbus_t *usbus, usbus_handler_t *handler);
+
+static usbus_dfu_state_t dfu_state;
+static uint32_t offset = 0;
+static uint8_t page_mem[FLASHPAGE_SIZE];
+static unsigned page_cpt;
+
+static size_t _gen_if_descriptor(usbus_t *usbus)
+{
+    usb_desc_if_dfu_t if_desc;
+
+    /* functional dfu descriptor */
+    if_desc.length = sizeof(usb_desc_if_dfu_t);
+    if_desc.type = USB_IF_DESCRIPTOR_DFU;
+    if_desc.attribute = USB_DFU_WILL_DETACH | USB_DFU_CAN_DOWNLOAD;
+    if_desc.detach_timeout = USB_DFU_DETACH_TIMEOUT_MS;
+    if_desc.xfer_size = FLASHPAGE_SIZE;
+    if_desc.bcd_dfu = USB_DFU_VERSION_BCD;
+
+    usbus_control_slicer_put_bytes(usbus, (uint8_t *)&if_desc, sizeof(if_desc));
+    return sizeof(usb_desc_if_dfu_t);
+}
+
+static size_t _gen_dfu_descriptor(usbus_t *usbus, void *arg)
+{
+    (void)arg;
+    size_t total_len = _gen_if_descriptor(usbus);
+    return total_len;
+}
+
+
+static const usbus_handler_driver_t dfu_driver = {
+    .init = _init,
+    .event_handler = _event_handler,
+    .transfer_handler = _transfer_handler,
+    .control_handler = _control_handler,
+};
+
+/* Descriptors */
+static const usbus_descr_gen_funcs_t _dfu_descriptor = {
+    .fmt_post_descriptor = _gen_dfu_descriptor,
+    .fmt_pre_descriptor = NULL,
+    .len = {
+        .fixed_len =  sizeof(usb_desc_if_dfu_t),
+    },
+    .len_type = USBUS_DESCR_LEN_FIXED,
+};
+
+static void _flashpage(void) {
+    /* reset offset value for next transfer */
+    offset = 0;
+    /* Erase page content */
+    flashpage_write(page_cpt, NULL);
+    /* Write buffer into flash */
+    int ret = flashpage_write_and_verify(page_cpt, page_mem);
```

Comment:
If you use the newly renamed `flashpage_write()` here you don't have to buffer a full page. But we can leave that for a follow up if you prefer.